File: jsync_js/packages/jsync/src/producer/digest.ts
import { blake3 } from '@noble/hashes/blake3.js';

import { JsyncError, JsyncErrorKind } from '../error.js';
import type { JsonValue } from '../value.js';

export function digestValue(value: JsonValue): string {
  const hasher = blake3.create();
  updateDigestValue(hasher, value);
  return bytesToHex(hasher.digest());
}

type ValueDigestHasher = ReturnType<typeof blake3.create>;

const DIGEST_TEXT_ENCODER = new TextEncoder();

function updateDigestValue(hasher: ValueDigestHasher, value: JsonValue): void {
  if (value === null) {
    hasher.update(Uint8Array.of(0x4e));
    return;
  }
  if (typeof value === 'boolean') {
    hasher.update(value ? Uint8Array.of(0x42, 0x31) : Uint8Array.of(0x42, 0x30));
    return;
  }
  if (typeof value === 'number') {
    updateDigestNumber(hasher, value);
    return;
  }
  if (typeof value === 'string') {
    hasher.update(Uint8Array.of(0x53));
    updateDigestBytes(hasher, DIGEST_TEXT_ENCODER.encode(value));
    return;
  }
  if (Array.isArray(value)) {
    hasher.update(Uint8Array.of(0x41));
    updateDigestLength(hasher, value.length);
    for (const child of value) {
      updateDigestValue(hasher, child);
    }
    return;
  }

  hasher.update(Uint8Array.of(0x4f));
  const keys = Object.keys(value).sort();
  updateDigestLength(hasher, keys.length);
  for (const key of keys) {
    hasher.update(Uint8Array.of(0x4b));
    updateDigestBytes(hasher, DIGEST_TEXT_ENCODER.encode(key));
    updateDigestValue(hasher, value[key]);
  }
}

function updateDigestNumber(hasher: ValueDigestHasher, value: number): void {
  if (!Number.isFinite(value)) {
    throw new JsyncError(
      JsyncErrorKind.InvalidJsonValue,
      'A non-finite number is not allowed in JSON.',
    );
  }
  if (Number.isInteger(value)) {
    if (!Number.isSafeInteger(value)) {
      throw new JsyncError(
        JsyncErrorKind.InvalidJsonValue,
        'The JSON integer is outside the cross-language safe integer range.',
      )
        .withMetadata('minimum', Number.MIN_SAFE_INTEGER)
        .withMetadata('maximum', Number.MAX_SAFE_INTEGER)
        .withMetadata('value', value);
    }
    updateDigestInteger(hasher, value);
    return;
  }

  hasher.update(Uint8Array.of(0x46));
  const bytes = new Uint8Array(8);
  new DataView(bytes.buffer).setFloat64(0, value, false);
  hasher.update(bytes);
}

function updateDigestInteger(hasher: ValueDigestHasher, value: number): void {
  hasher.update(Uint8Array.of(0x49));
  updateDigestBytes(hasher, DIGEST_TEXT_ENCODER.encode(value.toString()));
}

function updateDigestBytes(hasher: ValueDigestHasher, bytes: Uint8Array): void {
  updateDigestLength(hasher, bytes.length);
  hasher.update(bytes);
}

function updateDigestLength(hasher: ValueDigestHasher, length: number): void {
  const bytes = new Uint8Array(8);
  const view = new DataView(bytes.buffer);
  view.setUint32(0, Math.floor(length / 0x1_0000_0000), false);
  view.setUint32(4, length >>> 0, false);
  hasher.update(bytes);
}

function bytesToHex(bytes: Uint8Array): string {
  return Array.from(bytes, (byte) => byte.toString(16).padStart(2, '0')).join('');
}

File: jsync_rs/crates/jsync_rs/src/producer/digest.rs
use serde_json::{Number, Value};

use crate::error::{JsyncError, JsyncErrorKind};

pub(super) type ValueDigest = [u8; 32];

pub(super) fn digest_value(value: &Value) -> Result<ValueDigest, JsyncError> {
    let mut hasher = blake3::Hasher::new();
    update_digest_value(&mut hasher, value)?;
    Ok(hasher.finalize().into())
}

fn update_digest_value(hasher: &mut blake3::Hasher, value: &Value) -> Result<(), JsyncError> {
    match value {
        Value::Null => hasher.update(b"N"),
        Value::Bool(false) => hasher.update(b"B0"),
        Value::Bool(true) => hasher.update(b"B1"),
        Value::Number(number) => return update_digest_number(hasher, number),
        Value::String(value) => {
            hasher.update(b"S");
            update_digest_bytes(hasher, value.as_bytes())
        }
        Value::Array(values) => {
            hasher.update(b"A");
            update_digest_len(hasher, values.len());
            for value in values {
                update_digest_value(hasher, value)?;
            }
            hasher
        }
        Value::Object(object) => {
            hasher.update(b"O");
            update_digest_len(hasher, object.len());
            let mut keys = object.keys().collect::<Vec<_>>();
            keys.sort();
            for key in keys {
                hasher.update(b"K");
                update_digest_bytes(hasher, key.as_bytes());
                update_digest_value(hasher, &object[key])?;
            }
            hasher
        }
    };
    Ok(())
}

fn update_digest_number(hasher: &mut blake3::Hasher, number: &Number) -> Result<(), JsyncError> {
    if let Some(value) = number.as_i64() {
        validate_safe_json_integer(value as i128)?;
        update_digest_integer(hasher, value as i128);
        return Ok(());
    }
    if let Some(value) = number.as_u64() {
        validate_safe_json_integer(value as i128)?;
        update_digest_integer(hasher, value as i128);
        return Ok(());
    }

    let text = number.to_string();
    if !text.contains('.') && !text.contains('e') && !text.contains('E') {
        return Err(JsyncError::new(
            JsyncErrorKind::InvalidJsonValue,
            "The JSON integer is outside the supported CBOR integer range.",
        ));
    }
    if let Some(value) = number.as_f64() {
        validate_safe_json_float(value)?;
        if value.fract() == 0.0 {
            update_digest_integer(hasher, value as i128);
        } else {
            hasher.update(b"F");
            hasher.update(&value.to_be_bytes());
        }
        return Ok(());
    }

    Err(JsyncError::new(
        JsyncErrorKind::InvalidJsonValue,
        "The JSON number cannot be encoded as a CBOR number.",
    ))
}

fn update_digest_integer(hasher: &mut blake3::Hasher, value: i128) {
    hasher.update(b"I");
    update_digest_bytes(hasher, value.to_string().as_bytes());
}

fn update_digest_bytes<'a>(hasher: &'a mut blake3::Hasher, bytes: &[u8]) -> &'a mut blake3::Hasher {
    update_digest_len(hasher, bytes.len());
    hasher.update(bytes)
}

fn update_digest_len(hasher: &mut blake3::Hasher, len: usize) {
    hasher.update(&(len as u64).to_be_bytes());
}

fn validate_safe_json_integer(value: i128) -> Result<(), JsyncError> {
    const MAX_SAFE_JSON_INTEGER: i128 = 9_007_199_254_740_991;
    if (-MAX_SAFE_JSON_INTEGER..=MAX_SAFE_JSON_INTEGER).contains(&value) {
        Ok(())
    } else {
        Err(JsyncError::new(
            JsyncErrorKind::InvalidJsonValue,
            "The JSON integer is outside the cross-language safe integer range.",
        )
        .with_metadata("minimum", (-MAX_SAFE_JSON_INTEGER).to_string())
        .with_metadata("maximum", MAX_SAFE_JSON_INTEGER.to_string())
        .with_metadata("value", value.to_string()))
    }
}

fn validate_safe_json_float(value: f64) -> Result<(), JsyncError> {
    const MAX_SAFE_JSON_INTEGER: f64 = 9_007_199_254_740_991.0;
    if value.fract() != 0.0 || value.abs() <= MAX_SAFE_JSON_INTEGER {
        Ok(())
    } else {
        Err(JsyncError::new(
            JsyncErrorKind::InvalidJsonValue,
            "The JSON integer is outside the cross-language safe integer range.",
        )
        .with_metadata("minimum", (-MAX_SAFE_JSON_INTEGER).to_string())
        .with_metadata("maximum", MAX_SAFE_JSON_INTEGER.to_string())
        .with_metadata("value", value.to_string()))
    }
}
